packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt

```diff
@@ -68,33 +76,42 @@ internal object RealmObjectHelper {
         propertyName: String,
     ): Any? {
         obj.checkValid()
-        return realmValueToRealmObject(getValue(obj, propertyName), R::class, obj.mediator, obj.owner)
+        return realmValueToRealmObject(
+            getValue(obj, propertyName),
+            R::class,
+            obj.mediator,
+            obj.owner
+        )
     }
 
     // Return type should be RealmList<R?> but causes compilation errors for native
     @Suppress("unused") // Called from generated code
     internal inline fun <reified R : Any> getList(
         obj: RealmObjectReference<out BaseRealmObject>,
         propertyName: String
-    ): ManagedRealmList<Any?> = getList<R>(obj, propertyName, R::class)
-
-    internal fun <R> getList(
-        obj: RealmObjectReference<out BaseRealmObject>,
-        propertyName: String,
-        elementType: KClass<*>,
-    ): ManagedRealmList<Any?> =
-        getListByKey(obj, obj.propertyInfoOrThrow(propertyName).key, elementType)
+    ): ManagedRealmList<Any?> {
+        val elementType = R::class
+        val realmObjectCompanion = elementType.realmObjectCompanionOrNull()
+        return getListByKey(
+            obj,
+            obj.propertyInfoOrThrow(propertyName).key,
+            elementType,
+            realmObjectCompanion != null,
+        )
+    }
 
     // Cannot call managedRealmList directly from an inline function
     internal fun <R> getListByKey(
         obj: RealmObjectReference<out BaseRealmObject>,
         key: io.realm.internal.interop.PropertyKey,
         elementType: KClass<*>,
+        isObjectList: Boolean,
```

Comment:
It is a bit difficult to share as much code as possible without this dynamic path. The list operator requires the native pointer amongst others to avoid passing those around internally on all the list operations and this makes it a bit annoying to create the operator completely separate from the managed list. I have tried to extract building the operator into a separate method and inline some of the calls. Hopefully it helps. ... until we also need to take embeddedness into account 🤪